Zend/zend_object_handlers.c

```diff
@@ -1026,6 +1078,12 @@ ZEND_API void zend_std_unset_property(zval *object, zval *member, void **cache_s
 		zval *slot = OBJ_PROP(zobj, property_offset);
 
 		if (Z_TYPE_P(slot) != IS_UNDEF) {
+			if (UNEXPECTED(Z_ISREF_P(slot)) && Z_REFCOUNT_P(slot) > 1) {
+				zend_property_info *prop_info = zend_hash_find_ptr(&zobj->ce->properties_info, name);
```

Comment:
This should probably also use `zend_object_fetch_property_type_info`?

Reply:
Not just as an optimization -- directly looking up the property info does not correctly handle shadowing:

```php
class A {
    private int $prop = 42;

    public function test() {
        $x =& $this->prop;
        unset($this->prop);
        $x = "foo";
        var_dump($x);
    }
}
class B extends A {
    private $prop;
}

$b = new B;
$b->test();
```

Reply:
Here we also have the same issue as when destroying the object - it will always attempt to remove the prop_info of the _current_ class. If there has been a visibility check and the parents protected prop_info has been stored in the type sources set, we're fucked here.

We really should compare zend_object * and property_offset here.